File: mpiprog/clang/gather.c
/* File: gather.c
 * Name: D.Saravanan
 * Date: 26/11/2023
 * Program that receives data to process 0 from all other processes in a communicator
*/

#include <stdio.h>
#include <stdlib.h>
#include <math.h>
#include <mpi.h>

int main(int argc, char **argv) {

	int rank, size;

	MPI_Init(&argc, &argv);
	MPI_Comm_rank(MPI_COMM_WORLD, &rank);
	MPI_Comm_size(MPI_COMM_WORLD, &size);

	double time = MPI_Wtime();

	int N = 200; /* array size */
	double *data = (rank == 0) ? (double *) calloc(N, sizeof(data)) : NULL;

	if (rank == 0) /* initialize array if rank = 0 */
		for (size_t i = 0; i < N; data[i] = i * 1.0, i++);

	int workload[size];
	for (size_t i = 0; i < size; workload[i] = N / size, i++);
	for (size_t i = 0; i < N % size; workload[i] += 1, i++);

	size_t n = workload[rank]; /* number of elements sent to each process */
	double *ndata = (double *) calloc(n, sizeof(*ndata));

	/* send data from process 0 to all other processes in a communicator */
	MPI_Scatter(data, n, MPI_DOUBLE, ndata, n, MPI_DOUBLE, 0, MPI_COMM_WORLD);

	/* assign values */
	for (size_t i = 0; i < n; ndata[i] = 2 * ndata[i], i++);

	printf("Scatter data (process = %d):\n", rank);
	for (size_t i = n - 5; i < n; i++)
		printf("rank = %d   data[%d] = %0.1lf\n", rank, i, ndata[i]);

	/* receive data to process 0 from all other processes in a communicator */
	MPI_Gather(ndata, n, MPI_DOUBLE, data, n, MPI_DOUBLE, 0, MPI_COMM_WORLD);

	free(ndata);

	if (rank == 0) {
	printf("Gather data (process = 0):\n");
	for (size_t i = N - 5; i < N; i++)
		printf("rank = %d   data[%d] = %0.1lf\n", rank, i, data[i]);
	}

	free(data);

	time = MPI_Wtime() - time;

	printf("Timing from rank %d is %lf seconds\n", rank, time);

	MPI_Finalize();
	return 0;
}

File: mpiprog/python/gather.py
#!/usr/bin/env python
# File: gather.py
# Name: D.Saravanan
# Date: 26/11/2023

""" Script that receives data to process 0 from all other processes in a communicator """

import numpy as np
from mpi4py import MPI


def main():
    """gathering numpy arrays"""

    comm = MPI.COMM_WORLD
    rank = comm.Get_rank()
    size = comm.Get_size()

    time = MPI.Wtime()

    N = 200  # array size

    # initialize array if rank = 0
    data = np.array([i * 1.0 for i in range(0, N)]) if rank == 0 else None

    workload = [N // size for _ in range(size)]
    for n in range(N % size):
        workload[n] += 1

    n = workload[rank]  # number of elements send to each process
    ndata = np.zeros(workload[rank])

    # send data from process 0 to all other processes in a communicator
    comm.Scatter(data, ndata, root=0)

    # assign values
    ndata = 2 * ndata

    print(f"Scatter data (process = {rank}):")
    for i in range(n - 5, n):
        print(f"rank = {rank}   data[{i}] = {ndata[i]}")

    # receive data to process 0 from all other processes in a communicator
    comm.Gather(ndata, data, root=0)

    del ndata

    if rank == 0:
        print("Gather data (process = 0)")
        for i in range(N - 5, N):
            print(f"rank = {rank}   data[{i}] = {data[i]}")

    del data

    time = MPI.Wtime() - time

    print(f"Timing from rank {rank} is {time:0.6f} seconds")


if __name__ == "__main__":
    main()
